README.md
# LeadConduit Standard Integration

This module is for use on the [LeadConduit](http://activeprospect.com/products/leadconduit/) platform. Please see the [license agreement](http://creativecommons.org/licenses/by-nc-nd/4.0/)

[![Build Status](https://github.com/activeprospect/leadconduit-integration-standard/workflows/Node.js%20CI/badge.svg)](https://github.com/activeprospect/leadconduit-integration-standard/actions)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.2.3]
### Fixed
- Bump version of `leadconduit-default` to fix verbose response for JSON with numeric ID ([ch23913](https://app.clubhouse.io/active-prospect/story/23913/pipedrive-form-post-delivery))
- Add test to verify

## [1.2.1] - 2019-01-10
### Fixed
- Fix failing test that was missed due to dependency differences

## [1.2.0] - 2018-11-8
### Added
- Capture price variable

## [1.1.0] - 2018-02-26
### Added
- Add classic inbound

## [1.0.1] - 2018-02-19
### Added
- Convert to JS

## [1.0.0] - 2016-07-21
### Added
- Respond with HTTP 409 when feedback is forbidden

## [0.1.0] - 2016-07-20
### Added
- Feedback integration

## [0.0.1] - 2016-06-23
### Added
- Verbose inbound integration


test/inbound/classic-spec.js
const assert = require('chai').assert;
const integration = require('../../lib/inbound/classic');

describe('Classic Inbound Response', () => {

  beforeEach(() => {
    this.vars = {
      lead: { id: '123' },
      outcome: 'Failure',
      reason: 'bad!'
    };
  });


  it('should respond with 201', () => {
    const res = integration.response(baseRequest('application/json'), this.vars);
    assert.deepEqual(res.status, 201);
  });

  it('should respond with xml', () => {
    const res = integration.response(baseRequest('application/json'), this.vars);
    assert.deepEqual(res.headers, {'Content-Type': 'application/xml', 'Content-Length': 272});
  });

  it('should correctly parse failure outcome', () => {
    const res = integration.response(baseRequest('application/json'), this.vars);
    assert.deepEqual(res.body, nonSuccessBody);
  });

  it('should correctly parse success outcome', () => {
    this.vars.outcome = 'Success';
    this.vars.price = 1.5;
    delete this.vars.reason;

    const res = integration.response(baseRequest('application/json'), this.vars);
    assert.deepEqual(res.body, successBody);
  });
});


const baseRequest = (accept = null, querystring = '') => {
  return {
    uri: `/whatever${querystring}`,
    method: 'post',
    version: '1.1',
    headers: {
      'Accept': accept || 'application/xml',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: 'first_name=Joe',
    timestamp: new Date().getTime()
  };
};

const successBody = `<!DOCTYPE response SYSTEM "https://app.leadconduit.com/dtd/response-v2-basic.dtd">
<response>
  <result>Success</result>
  <leadId>123</leadId>
  <url>
    <![CDATA[https://app.leadconduit.com/leads?id=123]]>
  </url>
  <price>1.5</price>
</response>`;

const nonSuccessBody = `<!DOCTYPE response SYSTEM "https://app.leadconduit.com/dtd/response-v2-basic.dtd">
<response>
  <result>Failure</result>
  <reason>bad!</reason>
  <leadId>123</leadId>
  <url>
    <![CDATA[https://app.leadconduit.com/leads?id=123]]>
  </url>
  <price>0</price>
</response>`;


test/inbound/verbose-spec.js
const assert = require('chai').assert;
const integration = require('../../lib/inbound/verbose');
const types = require('leadconduit-types');

describe('Inbound Verbose Response', () => {

  beforeEach(() => {
    this.vars = {
      outcome: 'success',
      lead: {
        id: '1234'
      },
      appended: {
        briteverify: {
          email: {
            status: 'valid',
            disposable: 'false',
            role_address: 'false',
            outcome: 'success'
          }
        }
      },
      price: 1.5
    };
  });

  it('should set appended fields correctly in JSON response', () => {
    const req = {
      uri: 'http://example.com',
      headers: {
        'Accept': 'application/json'
      }
    };
    const expected = {
      status: 201,
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': 174
      },
      body: '{"appended":{"briteverify":{"email":{"status":"valid","disposable":"false","role_address":"false","outcome":"success"}}},"outcome":"success","lead":{"id":"1234"},"price":1.5}'
    };
    assert.deepEqual(integration.response(req, this.vars), expected);
  });

  it('should set appended fields correctly for JSON with a numeric ID', () => {
    const req = {
      uri: 'http://example.com',
      headers: {
        'Accept': 'application/json'
      }
    };

    this.vars.appended.otherservice = {
      user: {
        12: {
          foo: "bar"
        }
      }
    };

    const expected = {
      status: 201,
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': 174
      },
      body: '{"appended":{"briteverify":{"email":{"status":"valid","disposable":"false","role_address":"false","outcome":"success"}},"otherservice":{"user":{"\'12\'":{"foo":"bar"}}}},"outcome":"success","lead":{"id":"1234"},"price":1.5}'
    };
    assert.equal(integration.response(req, this.vars).body, expected.body);
  });

  it('should set appended fields correctly in XML response', () => {
    const req = {
      uri: 'http://example.com',
      headers: {
        'Accept': 'application/xml'
      }
    };
    const expected = {
      status: 201,
      headers: {
        'Content-Type': 'application/xml',
        'Content-Length': 368
      },
      body: '<?xml version="1.0"?>\n<result>\n  <appended>\n    <briteverify>\n      <email>\n        <status>valid</status>\n        <disposable>false</disposable>\n        <role_address>false</role_address>\n        <outcome>success</outcome>\n      </email>\n    </briteverify>\n  </appended>\n  <outcome>success</outcome>\n  <lead>\n    <id>1234</id>\n  </lead>\n  <price>1.5</price>\n</result>'
    };
    assert.deepEqual(integration.response(req, this.vars), expected);
  });


  it('should respond correctly when no fields are appended', () => {
    const req = {
      uri: 'http://example.com',
      headers: {
        'Accept': 'application/json'
      }
    };
    const vars = {
      outcome: 'success',
      lead: {
        id: '1234'
      }
    };
    const expected = {
      status: 201,
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': 52
      },
      body: '{"outcome":"success","lead":{"id":"1234"},"price":0}'
    };
    assert.deepEqual(integration.response(req, vars), expected);
  });


  it('should respond correctly with rich appended fields', () => {
    const req = {
      uri: 'http://example.com',
      headers: {
        'Accept': 'application/xml'
      }
    };
    const vars = {
      outcome: 'success',
      lead: {
        id: '1234'
      },
      price: 1.5,
      appended: {
        briteverify: {
          email: {
            status: 'valid',
            disposable: types.boolean.parse('false'),
            role_address: types.boolean.parse('false'),
            outcome: 'success',
            billable: types.number.parse('1')
          }
        }
      }
    };
    const expected = {
      status: 201,
      headers: {
        'Content-Type': 'application/xml',
        'Content-Length': 399
      },
      body: '<?xml version="1.0"?>\n<result>\n  <appended>\n    <briteverify>\n      <email>\n        <status>valid</status>\n        <disposable>false</disposable>\n        <role_address>false</role_address>\n        <outcome>success</outcome>\n        <billable>1</billable>\n      </email>\n    </briteverify>\n  </appended>\n  <outcome>success</outcome>\n  <lead>\n    <id>1234</id>\n  </lead>\n  <price>1.5</price>\n</result>'
    };
    assert.deepEqual(integration.response(req, vars), expected);
  });
});


test/inbound/feedback-spec.js
const assert = require('chai').assert;
const integration = require('../../lib/inbound/feedback');



describe('Inbound feedback', () => {

  describe('request', () => {

    it('should not allow head', () => {
      assertMethodNotAllowed('head');
    });

    it('should not allow put', () => {
      assertMethodNotAllowed('put');
    });

    it('should not allow delete', () => {
      assertMethodNotAllowed('delete');
    });

    it('should not allow patch', () => {
      assertMethodNotAllowed('patch');
    });

    it('should handle GET', () => {
      const request = {
        uri: 'https://app.leadconduit.com/feedback?event_id=12345&type=return&reason=Wrong+number',
        method: 'GET',
        headers: {
          'Accept': 'application/json'
        }
      };
      assert.deepEqual(integration.request(request), {
        type: 'return',
        reason: 'Wrong number'
      });
    });

    it('should handle form POST', () => {
      const body = 'type=return&reason=Wrong+number';
      const request = {
        uri: 'https://app.leadconduit.com/feedback?event_id=12345',
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Accept': 'application/json',
          'Content-Length': Buffer.byteLength(body)
        },
        body: body
      };
      assert.deepEqual(integration.request(request),{
        type: 'return',
        reason: 'Wrong number'
      });
    });

    it('should handle XML', () => {
      const body = `
        <feedback>
          <type>return</type>
          <reason>Wrong number</reason>
        </feedback>
        `;
      const request = {
        uri: 'https://app.leadconduit.com/feedback?event_id=12345',
        method: 'POST',
        headers: {
          'Content-Type': 'text/xml',
          'Accept': 'text/xml',
          'Content-Length': Buffer.byteLength(body)
        },
        body: body
      };
      assert.deepEqual(integration.request(request),{
        type: 'return',
        reason: 'Wrong number'
      });
    });

    it('should handle JSON', () => {
      const body = `
        {
          "type": "return",
          "reason": "Wrong number"
        }`;
      const request = {
        uri: 'https://app.leadconduit.com/feedback?event_id=12345',
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'Content-Length': Buffer.byteLength(body)
        },
        body: body
      };
      assert.deepEqual(integration.request(request),{
        type: 'return',
        reason: 'Wrong number'
      });


      it('should handle malformed XML', () => {
        const body = 'whatever';
        const request = {
          uri: 'https://app.leadconduit.com/feedback?event_id=12345',
          method: 'POST',
          headers: {
            'Content-Type': 'text/xml',
            'Accept': 'text/xml',
            'Content-Length': Buffer.byteLength(body)
          },
          body: body
        };

        try {
          integration.request(request);
          assert.fail('expected an error to be thrown when xml content cannot be parsed');
        } catch (e) {
          assert.equal(e.status, 400);
          assert.equal(e.body, 'Body does not contain XML or XML is unparseable -- Error: Non-whitespace before first tag. Line: 0 Column: 1 Char: w.');
          assert.deepEqual(e.headers, {'Content-Type': 'text/plain'});
        }
      });


      it('should handle malformed JSON', () => {
        const body = 'whatever';

        const request = {
          uri: 'https://app.leadconduit.com/feedback?event_id=12345',
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'Content-Length': Buffer.byteLength(body)
          },
          body: body
        };

        try {
          integration.request(request);
          assert.fail('expected an error to be thrown when JSON content cannot be parsed');
        } catch (e) {
          assert.equal(e.status, 400);
          assert.equal(e.body, 'Body is not parsable as application/json -- Unexpected token w in JSON at position 0');
          assert.deepEqual(e.headers, {'Content-Type': 'text/plain'});
        }
      });
    });

    describe('response', () => {

      let vars = null;

      beforeEach(() => {
        vars = {
          lead: {
            id: '123',
            email: 'foo@bar.com'
          },
          price: 1.5,
          outcome: 'failure',
          reason: 'bad!'
        };
      });

      it('should respond with json', () => {
        const res = integration.response(baseRequest('application/json'), vars);
        assert.equal(res.status, 201);
        assert.deepEqual(res.headers, {'Content-Type': 'application/json', 'Content-Length': 92});
        assert.equal(res.body, '{"outcome":"failure","reason":"bad!","lead":{"id":"123","email":"foo@bar.com"},"price":-1.5}');
      });

      it('should default to json', () => {
        const res = integration.response(baseRequest('*/*'), vars);
        assert.equal(res.status, 201);
        assert.deepEqual(res.headers['Content-Type'],  'application/json');
      });

      it('should return 409 when feedback is disabled', () => {
        vars.reason = 'Feedback is forbidden';
        const res = integration.response(baseRequest('*/*'), vars);
        assert.equal(res.status, 409);
      });

      it('should respond with text xml', () => {
        const res = integration.response(baseRequest('text/xml'), vars);
        assert.equal(res.status, 201);
        assert.deepEqual(res.headers, {'Content-Type': 'text/xml', 'Content-Length': 232});
        assert.equal(res.body, '<?xml version="1.0"?>\n<result>\n  <outcome>failure</outcome>\n  <reason>bad!</reason>\n  <lead>\n    <id>123</id>\n    <first_name/>\n    <last_name/>\n    <email>foo@bar.com</email>\n    <phone_1/>\n  </lead>\n  <price>-1.5</price>\n</result>');
      });

      it('should respond with application xml', () => {
        const res = integration.response(baseRequest(), vars);
        assert.equal(res.status, 201);
        assert.deepEqual(res.headers, {'Content-Type': 'application/xml', 'Content-Length': 232});
        assert.equal(res.body, '<?xml version="1.0"?>\n<result>\n  <outcome>failure</outcome>\n  <reason>bad!</reason>\n  <lead>\n    <id>123</id>\n    <first_name/>\n    <last_name/>\n    <email>foo@bar.com</email>\n    <phone_1/>\n  </lead>\n  <price>-1.5</price>\n</result>');
      });
    });

    describe('With specified fields in response', () => {

      let vars = {};

      beforeEach(() => {
        vars = {
          lead: {
            id: '123',
            email: 'foo@bar.com'
          },
          outcome: 'failure',
          reason: 'bad!'
        };
      });

      it('should respond with json', () => {
        const res = integration.response(baseRequest('application/json'), vars, ['outcome', 'lead.id', 'lead.email']);
        assert.equal(res.status, 201);
        assert.equal(res.headers['Content-Type'], 'application/json');
        assert.equal(res.body, '{"outcome":"failure","lead":{"id":"123","email":"foo@bar.com"},"price":0}');
      });

      it('should respond with text xml', () => {
        const res = integration.response(baseRequest('text/xml'), vars, ['outcome', 'lead.id', 'lead.email']);
        assert.equal(res.status, 201);
        assert.equal(res.headers['Content-Type'], 'text/xml');
        assert.equal(res.body, '<?xml version="1.0"?>\n<result>\n  <outcome>failure</outcome>\n  <lead>\n    <id>123</id>\n    <email>foo@bar.com</email>\n  </lead>\n  <price>0</price>\n</result>');
      });

      it('should respond with application xml', () => {
        const res = integration.response(baseRequest(), vars, ['outcome', 'lead.id', 'lead.email']);
        assert.equal(res.status, 201);
        assert.equal(res.headers['Content-Type'], 'application/xml');
        assert.equal(res.body, '<?xml version="1.0"?>\n<result>\n  <outcome>failure</outcome>\n  <lead>\n    <id>123</id>\n    <email>foo@bar.com</email>\n  </lead>\n  <price>0</price>\n</result>');
      });
    });
  });
});

const baseRequest = (accept = null, querystring = '') => {
  return {
    uri: `/whatever${querystring}`,
    method: 'post',
    version: '1.1',
    headers: {
      'Accept': accept || 'application/xml',
      'Content-Type': 'application/x-www-form-urlencoded'
    },
    body: 'first_name=Joe',
    timestamp: new Date().getTime()
  };
};

const assertMethodNotAllowed = (method) => {
  try {
    integration.request({method: method});
    assert.fail(`expected ${method} to throw an error`);
  } catch (e) {
    assert.equal(e.status, 415);
    assert.equal(e.body, `The ${method.toUpperCase()} method is not allowed`);
    assert.deepEqual(e.headers,{
      'Allow': 'GET, POST',
      'Content-Type': 'text/plain'
    });
  }
};


lib/inbound/classic.js
const xmlbuilder = require('xmlbuilder');
const default_inbound = require('leadconduit-default').inbound;

const response = (req, vars) => {
  const body = buildXml(vars);

  return {
    status: 201,
    headers: {
      'Content-Type': 'application/xml',
      'Content-Length': body.length
    },
    body
  };
};

response.variables = default_inbound.response.variables;

module.exports = {
  name: 'Standard Classic',
  request: default_inbound.request,
  response
};

// This handles the logic around creating the reason tag
const buildXml = (vars) => {

  const url = `https://app.leadconduit.com/leads?id=${vars.lead.id}`;
  const price = vars.price || 0;
  const xml = xmlbuilder
    .create('response', {headless: true})
    .dtd('https://app.leadconduit.com/dtd/response-v2-basic.dtd').up()
    .element('result', vars.outcome).up();
    
  if (vars.reason) { xml.ele('reason', vars.reason); }
  xml.element('leadId', vars.lead.id).up()
    .element('url').dat(url).up()
    .element('price', price);

  return xml.end({pretty: true});
};


lib/outbound/to_flow.js
url = require('url');
flatten = require('flat');
querystring = require('querystring');

const request = (vars) => {

  let body = flatten(vars.lead);


  for(let key in body) {
    let value = body[key];
    if(value) body[key] = value.toString();
  }

  for(let key in vars.custom) {
    let value = vars.custom[key];
    if(value) body[key] = value;
  }

  body = querystring.stringify(body);

  return {
    body: body,
    url: vars.url,
    headers: {
      'Content-Type': 'application/json'
    }
  }

};

request.variables = () => {
  return [
    { name: 'url', type: 'string', required: true, description: '' },
    { name: 'custom.*', type: 'wildcard' }
  ];
};

const response = (vars, req, res) => {

  if(!res || !res.body) {
    return {
      outcome: 'error',
      reason: 'unknown response'
    }
  }

  try {
    let parsed = JSON.parse(res.body);

    return {
      outcome: parsed.outcome,
      reason: parsed.reason,
      lead: {
        id: parsed.lead.id
      }
    };
  }
  catch (e) {
    return {
      outcome: 'error',
      reason: 'unable to parse response'
    }
  }

};

response.variables = () => {
  return [
    { name: 'outcome', type: 'string', description: 'The outcome of the lead' },
    { name: 'reason', type: 'string', description: 'If the outcome was not success, this is the reason' },
    { name: 'lead.id', type: 'string', description: `The lead's ID` }
  ]
};

const validate = (vars) => {

  if (!vars.url) return 'URL must not be blank';
  let parsed = url.parse(vars.url);
  if (!(parsed.host && parsed.host.match(/leadconduit/))) return 'URL must be a valid LeadConduit posting URL';

};

module.exports = {
  name: 'To Flow',
  request: request,
  response: response,
  validate: validate
};

lib/inbound/feedback.js
const _ = require('lodash');
const url = require('url');
const flat = require('flat');
const mimecontent = require('mime-content');
const mimeparse = require('mimeparse');
const xmlbuilder = require('xmlbuilder');
const dotaccess = require('dotaccess');
const HttpError = require('leadconduit-integration').HttpError;

const supportedMimeTypes = [
  'application/x-www-form-urlencoded',
  'application/json',
  'application/xml',
  'text/xml'
];

const supportedMimeTypeLookup = supportedMimeTypes.reduce(((lookup, mimeType) => {
  lookup[mimeType] = true;
  return lookup;
}), {});


const request = (req) => {

  // ensure supported method
  const method = (req.method) ? req.method.toLowerCase() : null;
  if (method !== 'get' && method !== 'post') {
    throw new HttpError(415, { 'Content-Type': 'text/plain', Allow: 'GET, POST' }, `The ${method.toUpperCase()} method is not allowed`);
  }

  // ensure acceptable content type, preferring JSON
  const mimeType = selectMimeType(req.headers['Accept']);
  if (!mimeType) {
    throw new HttpError(406, { 'Content-Type': 'text/plain' }, 'Not capable of generating content according to the Accept header');
  }

  const feedbackVars = parseFeedbackVars(req);
  delete feedbackVars.event_id;
  return feedbackVars;
};


const parseFeedbackVars = (req) => {
  const method = (req.method) ? req.method.toLowerCase() : null;

  // parse the query string
  const uri = url.parse(req.uri, true);
  const query = flat.unflatten(uri.query);

  if (method === 'get') {
    return query;
  } else if (method === 'post') {

    if (req.headers['Content-Length'] || req.headers['Transfer-Encoding'] === 'chunked') {

      // assume a request body

      // ensure a content type header
      const contentType = req.headers['Content-Type'];
      if (!contentType) {
        throw new HttpError(415, {'Content-Type': 'text/plain'}, 'Content-Type header is required');
      }

      // ensure valid mime type
      const mimeType = selectMimeType(req.headers['Content-Type']);
      if (!supportedMimeTypeLookup[mimeType]) {
        throw new HttpError(406, {'Content-Type': 'text/plain'}, `MIME type in Content-Type header is not supported. Use only ${supportedMimeTypes.join(', ')}.`);
      }

      // parse request body according the the mime type
      const body = (req.body) ? req.body.trim() : null;
      if (!body) {
        return query;
      }

      let parsed;
      try {
        parsed = mimecontent(body, mimeType);
      } catch(e)  {
        throw new HttpError(400, { 'Content-Type': 'text/plain' }, `Body is not parsable as ${mimeType} -- ${e.message}`);
      }

      // if form URL encoding, convert dot notation keys
      if (mimeType == 'application/x-www-form-urlencoded') {
        parsed = flat.unflatten(parsed);
      }

      // if XML, turn doc into an object
      if (mimeType === 'application/xml' || mimeType === 'text/xml') {
        try {
          parsed = parsed.toObject({explicitArray: false, explicitRoot: false, mergeAttrs: true});
        } catch (e) {
          const xmlError = e.toString().replace(/\r?\n/g, ' ');
          throw new HttpError(400, {'Content-Type': 'text/plain'}, `Body does not contain XML or XML is unparseable -- ${xmlError}.`);
        }
      }

      // merge query string data into data parsed from request body
      _.merge(parsed, query);

      return parsed;
    } else {
      // assume no request body
      return query;
    }
  }
};


request.variables = () => [
  { name: 'type', label: 'Feedback type', type: 'string', description: 'The type of feedback being given', examples: ['return', 'conversion'] },
  { name: 'reason', label: 'Feedback reason', type: 'string', description: 'The reason the feedback is being given', examples: ['Disconnected phone', 'Wrong number', 'Uncontactable', 'New customer']}
];

const response = (req, vars, fieldIds = ['outcome', 'reason', 'lead.id', 'lead.first_name', 'lead.last_name', 'lead.email', 'lead.phone_1']) => {

  
  const price = vars.price || 0;
  const mimeType = selectMimeType(req.headers['Accept']);
  
  let rtn = {};
  for (let field of fieldIds) {
    rtn[field] = (dotaccess.get(vars, field)) ?
      dotaccess.get(vars, field).valueOf() :
      undefined;
  }
  rtn = flat.unflatten(rtn);

  // if outcome is failure, return negative price representing refund
  if (rtn.outcome === 'failure') {
    rtn.price = 0 - price;
  } else {
    rtn.price = price;
  }

  let body = '';
  if (mimeType === 'application/xml' || mimeType === 'text/xml') {
    body = xmlbuilder.create({result: rtn}).end({pretty: true});
  } else if (mimeType === 'application/json') {
    body = JSON.stringify(rtn);
  } else {
    const lines = [];
    for (let [fieldName, fieldValue] of new Map(Object.entries(rtn))) {
      lines.push(`${fieldName.replace(/./g, '_')}:${fieldValue || ''}`);
    }
    body = lines.join('\n');
  }

  const headers = {
    'Content-Type': mimeType,
    'Content-Length': Buffer.byteLength(body)
  };

  const status = (rtn.reason && rtn.reason.match(/forbidden/i)) ? 409 : 201;

  return {
    status: status,
    headers: headers,
    body: body
  };
};


response.variables = () => [
  { name: 'outcome', type: 'string', description: 'The outcome of the feedback request (default is success, meaning that the feedback was accepted)' },
  { name: 'reason', type: 'string', description: 'If the outcome was a failure, this is the reason' },
  { name: 'price', type: 'number', description: 'The price of the lead' },
  { name: 'lead.id', type: 'string', description: 'The lead identifier' },
  { name: 'lead.first_name', type: 'string', description: 'The consumer\'s first name' },
  { name: 'lead.last_name', type: 'string', description: 'The consumer\'s last name' },
  { name: 'lead.email', type: 'string', description: 'The consumer\'s email address' },
  { name: 'lead.phone_1', type: 'string', description: 'The consumer\'s phone number' }
];


const selectMimeType = (contentType) => {
  contentType = contentType || 'application/json';
  if (contentType === '*/*') contentType = 'application/json';
  return mimeparse.bestMatch(supportedMimeTypes, contentType);
};

module.exports = {
  name: 'Standard Feedback',
  request,
  response
};
